{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "15f828ca",
   "metadata": {},
   "source": [
    "### Goal - Notebook Description\n",
    "\n",
    "\n",
    "- Read gene expression data from MelanoDB\n",
    "- Create precessed Table \n",
    "- Create simple vizualisation (heatmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c688678f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "import time\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import json\n",
    "import sqlite3\n",
    "import os\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "import ast\n",
    "from joblib import Parallel, delayed\n",
    "import multiprocessing as mp\n",
    "from multiprocessing import Pool\n",
    "from functools import partial\n",
    "import get_ge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "362cc279",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load MelanoDB\n",
    "cnx = sqlite3.connect('../database/MelanoDB.db')\n",
    "df_ge =  pd.read_sql_query(\"SELECT * FROM gene_expressions\", cnx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "764de8df",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4466466\n"
     ]
    }
   ],
   "source": [
    "df_ge = df_ge[df_ge['temporality'] == 'pre treatment']\n",
    "#df_ge = df_ge[df_ge['value'] != 0]\n",
    "print(len(df_ge))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2dab5fb6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{\"{'title': 'Acquired resistance and clonal evolution in melanoma during BRAF inhibitor therapy', 'author': 'Hugo W., Roger S. Lo', 'journal': 'Cancers Discovery', 'location': 'Los Angeles (United State)', 'date': 2014}\",\n",
       " \"{'title': 'BRAF Inhibitor Resistance Mechanisms in Metastatic Melanoma: Spectrum and Clinical Impact', 'author': 'Helen Rizos, Georgina V.Long', 'journal': 'Clinical Cancer Research', 'location': 'Sydney (Australia)', 'date': 2014}\",\n",
       " \"{'title': 'Baseline Genomic Features in BRAFV600-Mutated Metastatic Melanoma Patients Treated with BRAF Inhibitor + MEK Inhibitor in Routine Care', 'author': 'Baptiste Louveau, Samia Mourah', 'journal': 'Cancers (Basel)', 'location': 'Paris (France)', 'date': 2019}\",\n",
       " \"{'title': 'Co-clinical assessment identifies patterns of BRAF inhibitor resistance in melanoma', 'author': 'Lawrence N. Kwong, Lynda Chin', 'journal': 'Journal of Clinical Investigations ', 'location': 'Boston (United States)', 'date': 2015}\",\n",
       " \"{'title': 'Genomic Features of Exceptional Response in Vemurafenib + Cobimetinib–treated Patients with BRAFV600-mutated Metastatic Melanoma', 'author': 'Yibing Yan, Antoni Ribas', 'journal': 'Clinical Cancer Research', 'location': 'San Francisco, California (United States)', 'date': 2019}\",\n",
       " \"{'title': 'Increased MAPK reactivation in early resistance to dabrafenib/trametinib combination therapy of BRAF-mutant metastatic melanoma', 'author': 'Georgina V.Long, Helen Rizos', 'journal': 'Nature Communications', 'location': 'Sydney (Australia)', 'date': 2014}\"}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(list(df_ge['source']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a4716d6a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2961950\n",
      "{\"{'title': 'Co-clinical assessment identifies patterns of BRAF inhibitor resistance in melanoma', 'author': 'Lawrence N. Kwong, Lynda Chin', 'journal': 'Journal of Clinical Investigations ', 'location': 'Boston (United States)', 'date': 2015}\", \"{'title': 'Acquired resistance and clonal evolution in melanoma during BRAF inhibitor therapy', 'author': 'Hugo W., Roger S. Lo', 'journal': 'Cancers Discovery', 'location': 'Los Angeles (United State)', 'date': 2014}\", \"{'title': 'Genomic Features of Exceptional Response in Vemurafenib + Cobimetinib–treated Patients with BRAFV600-mutated Metastatic Melanoma', 'author': 'Yibing Yan, Antoni Ribas', 'journal': 'Clinical Cancer Research', 'location': 'San Francisco, California (United States)', 'date': 2019}\"}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>id</th>\n",
       "      <th>creation_datetime</th>\n",
       "      <th>patientID</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>HGNC</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>description</th>\n",
       "      <th>value</th>\n",
       "      <th>temporality</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>580</td>\n",
       "      <td>581</td>\n",
       "      <td>2024-07-10 11:47:53.298381</td>\n",
       "      <td>None</td>\n",
       "      <td>03660196C</td>\n",
       "      <td>None</td>\n",
       "      <td>GeneID:1</td>\n",
       "      <td>None</td>\n",
       "      <td>6.898152</td>\n",
       "      <td>pre treatment</td>\n",
       "      <td>{'title': 'Genomic Features of Exceptional Res...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>581</td>\n",
       "      <td>582</td>\n",
       "      <td>2024-07-10 11:47:53.298381</td>\n",
       "      <td>None</td>\n",
       "      <td>03660196C</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>GeneID:10</td>\n",
       "      <td>None</td>\n",
       "      <td>0.068518</td>\n",
       "      <td>pre treatment</td>\n",
       "      <td>{'title': 'Genomic Features of Exceptional Res...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>582</td>\n",
       "      <td>583</td>\n",
       "      <td>2024-07-10 11:47:53.298381</td>\n",
       "      <td>None</td>\n",
       "      <td>03660196C</td>\n",
       "      <td>ADA</td>\n",
       "      <td>GeneID:100</td>\n",
       "      <td>None</td>\n",
       "      <td>30.686816</td>\n",
       "      <td>pre treatment</td>\n",
       "      <td>{'title': 'Genomic Features of Exceptional Res...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>583</td>\n",
       "      <td>584</td>\n",
       "      <td>2024-07-10 11:47:53.298381</td>\n",
       "      <td>None</td>\n",
       "      <td>03660196C</td>\n",
       "      <td>CDH2</td>\n",
       "      <td>GeneID:1000</td>\n",
       "      <td>None</td>\n",
       "      <td>0.180316</td>\n",
       "      <td>pre treatment</td>\n",
       "      <td>{'title': 'Genomic Features of Exceptional Res...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>584</td>\n",
       "      <td>585</td>\n",
       "      <td>2024-07-10 11:47:53.298381</td>\n",
       "      <td>None</td>\n",
       "      <td>03660196C</td>\n",
       "      <td>AKT3</td>\n",
       "      <td>GeneID:10000</td>\n",
       "      <td>None</td>\n",
       "      <td>17.434889</td>\n",
       "      <td>pre treatment</td>\n",
       "      <td>{'title': 'Genomic Features of Exceptional Res...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index   id           creation_datetime patientID  sample_id  HGNC  \\\n",
       "0    580  581  2024-07-10 11:47:53.298381      None  03660196C  None   \n",
       "1    581  582  2024-07-10 11:47:53.298381      None  03660196C  NAT2   \n",
       "2    582  583  2024-07-10 11:47:53.298381      None  03660196C   ADA   \n",
       "3    583  584  2024-07-10 11:47:53.298381      None  03660196C  CDH2   \n",
       "4    584  585  2024-07-10 11:47:53.298381      None  03660196C  AKT3   \n",
       "\n",
       "         GeneID description      value    temporality  \\\n",
       "0      GeneID:1        None   6.898152  pre treatment   \n",
       "1     GeneID:10        None   0.068518  pre treatment   \n",
       "2    GeneID:100        None  30.686816  pre treatment   \n",
       "3   GeneID:1000        None   0.180316  pre treatment   \n",
       "4  GeneID:10000        None  17.434889  pre treatment   \n",
       "\n",
       "                                              source  \n",
       "0  {'title': 'Genomic Features of Exceptional Res...  \n",
       "1  {'title': 'Genomic Features of Exceptional Res...  \n",
       "2  {'title': 'Genomic Features of Exceptional Res...  \n",
       "3  {'title': 'Genomic Features of Exceptional Res...  \n",
       "4  {'title': 'Genomic Features of Exceptional Res...  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Filter regarding sources\n",
    "sources_to_keep = [\"{'title': 'Acquired resistance and clonal evolution in melanoma during BRAF inhibitor therapy', 'author': 'Hugo W., Roger S. Lo', 'journal': 'Cancers Discovery', 'location': 'Los Angeles (United State)', 'date': 2014}\",\n",
    "                  \"{'title': 'Co-clinical assessment identifies patterns of BRAF inhibitor resistance in melanoma', 'author': 'Lawrence N. Kwong, Lynda Chin', 'journal': 'Journal of Clinical Investigations ', 'location': 'Boston (United States)', 'date': 2015}\",\n",
    "                  \"{'title': 'Genomic Features of Exceptional Response in Vemurafenib + Cobimetinib–treated Patients with BRAFV600-mutated Metastatic Melanoma', 'author': 'Yibing Yan, Antoni Ribas', 'journal': 'Clinical Cancer Research', 'location': 'San Francisco, California (United States)', 'date': 2019}\"]\n",
    "df_ge = df_ge[df_ge['source'].isin(sources_to_keep)]\n",
    "df_ge.reset_index(inplace=True)\n",
    "print(len(df_ge))\n",
    "print(set(df_ge['source']))\n",
    "df_ge.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "c73b696c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def intersection(lst1, lst2, lst3):\n",
    "    lst4 = [value for value in lst2 if value in lst3]\n",
    "    intersect_all = [value for value in lst1 if value in lst4]\n",
    "    return lst4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e018e645",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get list of common genes\n",
    "list_genes_shi = list(set(df_ge[df_ge['source']==sources_to_keep[0]].HGNC))\n",
    "list_genes_kwong = list(set(df_ge[df_ge['source']==sources_to_keep[1]].HGNC))\n",
    "list_genes_ribas = list(set(df_ge[df_ge['source']==sources_to_keep[2]].HGNC))\n",
    "\n",
    "list_genes = intersection(list_genes_shi, list_genes_kwong, list_genes_ribas)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "351f3cec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filtering tab according to gene list\n",
    "df_ge = df_ge[df_ge['HGNC'].isin(list_genes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "5516b1d0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Patients Shi: 12\n",
      "# Patients kwong: 14\n",
      "# Patients ribas: 72\n"
     ]
    }
   ],
   "source": [
    "print('# Patients Shi: '+str(len(df_ge[df_ge['source']==sources_to_keep[0]].sample_id.value_counts())))\n",
    "print('# Patients kwong: ' +str(len(df_ge[df_ge['source']==sources_to_keep[1]].sample_id.value_counts())))\n",
    "print('# Patients ribas: '+str(len(df_ge[df_ge['source']==sources_to_keep[2]].sample_id.value_counts())))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4dc123b",
   "metadata": {},
   "source": [
    "## Get gene expression per sample - functions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "74d02138",
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_ge_df_value(tab_ge, df, samp, list_g):\n",
    "    print(samp)\n",
    "    for gene in list_g:\n",
    "        if (np.shape(df[(df['sample_id']==samp) & (df['HGNC']==gene)]) !=0):\n",
    "            if (np.any(df[(df['sample_id']==samp) & (df['HGNC']==gene)])):\n",
    "                val = df[(df['sample_id']==samp) & (df['HGNC']==gene)].value.to_numpy()[0]\n",
    "                print(val)\n",
    "                tab_ge.loc[gene, samp]=val\n",
    "    return(tab_ge)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "1f48cbe5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_ge_df_value_test(samp):\n",
    "    print(samp)\n",
    "    for gene in list_genes[0:5]:\n",
    "        if (np.shape(df_ge[(df_ge['sample_id']==samp) & (df_ge['HGNC']==gene)]) !=0):\n",
    "            if (np.any(df_ge[(df_ge['sample_id']==samp) & (df_ge['HGNC']==gene)])):\n",
    "                val = df_ge[(df_ge['sample_id']==samp) & (df_ge['HGNC']==gene)].value.to_numpy()[0]\n",
    "                print(val)\n",
    "                GE_tab_test.loc[gene, samp]=val\n",
    "    return(GE_tab_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "bc969fdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_ge_df_value_col(samp, df_ge, list_genes):\n",
    "    samp_tab = pd.DataFrame(index = list_genes[0:5])\n",
    "    #print(samp)\n",
    "    for gene in list_genes[0:5]:\n",
    "        if (np.shape(df_ge[(df_ge['sample_id']==samp) & (df_ge['HGNC']==gene)]) !=0):\n",
    "            if (np.any(df_ge[(df_ge['sample_id']==samp) & (df_ge['HGNC']==gene)])):\n",
    "                val = df_ge[(df_ge['sample_id']==samp) & (df_ge['HGNC']==gene)].value.to_numpy()[0]\n",
    "                print(val)\n",
    "                samp_tab.loc[gene, samp]=val\n",
    "    return(samp_tab)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "737e029e",
   "metadata": {},
   "source": [
    "## Get variance per genes & retrieve high variance gene list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "1353b364",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HGNC</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10727</th>\n",
       "      <td>MIR4649</td>\n",
       "      <td>2.096593e+10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10890</th>\n",
       "      <td>MIR5588</td>\n",
       "      <td>2.309566e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10700</th>\n",
       "      <td>MIR4517</td>\n",
       "      <td>9.855878e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10740</th>\n",
       "      <td>MIR4668</td>\n",
       "      <td>9.206484e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10644</th>\n",
       "      <td>MIR4442</td>\n",
       "      <td>5.202958e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8152</th>\n",
       "      <td>KRT6A</td>\n",
       "      <td>5.948337e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15378</th>\n",
       "      <td>RPS3A</td>\n",
       "      <td>5.849340e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15286</th>\n",
       "      <td>RPL21P28</td>\n",
       "      <td>5.796954e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8114</th>\n",
       "      <td>KRT10</td>\n",
       "      <td>5.620441e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15514</th>\n",
       "      <td>S100A1</td>\n",
       "      <td>5.540717e+05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           HGNC         value\n",
       "10727   MIR4649  2.096593e+10\n",
       "10890   MIR5588  2.309566e+09\n",
       "10700   MIR4517  9.855878e+08\n",
       "10740   MIR4668  9.206484e+08\n",
       "10644   MIR4442  5.202958e+08\n",
       "...         ...           ...\n",
       "8152      KRT6A  5.948337e+05\n",
       "15378     RPS3A  5.849340e+05\n",
       "15286  RPL21P28  5.796954e+05\n",
       "8114      KRT10  5.620441e+05\n",
       "15514    S100A1  5.540717e+05\n",
       "\n",
       "[100 rows x 2 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ge.groupby('HGNC', as_index=False)['value'].var().sort_values(by='value', ascending=False)[0:100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "c8c9764f",
   "metadata": {},
   "outputs": [],
   "source": [
    "high_variance_genes = list(df_ge.groupby('HGNC', as_index=False)['value'].var().sort_values(by='value', ascending=False)['HGNC'][0:500])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d20f86d9",
   "metadata": {},
   "source": [
    "## Get gene expression tab - Sequential iterations "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "da26e991",
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n",
      "2.026665692\n",
      "0.0\n",
      "0.007951394\n",
      "0.0\n",
      "0.136472765\n",
      "0.0\n",
      "0.056639186\n",
      "0.017666947\n",
      "0.134264596\n",
      "0.01\n",
      "0.38\n",
      "0.01\n",
      "1.64\n",
      "0.0\n",
      "0.53\n",
      "0.0\n",
      "0.122812043\n",
      "0.0\n",
      "0.08\n",
      "0.0\n",
      "0.967963937\n",
      "0.0\n",
      "0.037499961\n",
      "0.0\n",
      "1.54\n",
      "0.0\n",
      "0.037028924\n",
      "0.01597747\n",
      "0.11675478\n",
      "0.0\n",
      "0.15\n",
      "0.0\n",
      "0.101516511\n",
      "0.0\n",
      "0.246343348\n",
      "0.01784255\n",
      "0.139947\n",
      "0.0\n",
      "0.084368629\n",
      "0.01\n",
      "0.39\n",
      "0.0\n",
      "0.094403548\n",
      "0.0\n",
      "0.191821195\n",
      "0.0\n",
      "0.412164683\n",
      "0.0\n",
      "0.226420602\n",
      "0.0\n",
      "0.104606095\n",
      "0.01791285\n",
      "2.398925\n",
      "0.0\n",
      "0.372707888\n",
      "7.523176801\n",
      "8.307384777\n",
      "0.0\n",
      "0.837986152\n",
      "0.0\n",
      "2.066005036\n",
      "0.0\n",
      "0.075774843\n",
      "0.0\n",
      "0.053561446\n",
      "0.0\n",
      "6.29765\n",
      "0.0\n",
      "0.181347251\n",
      "0.0\n",
      "0.153302772\n",
      "0.0\n",
      "0.403671\n",
      "0.0\n",
      "0.454132981\n",
      "0.0\n",
      "0.7915855\n",
      "0.0\n",
      "0.108534678\n",
      "0.0\n",
      "0.643877937\n",
      "0.0\n",
      "1.822334345\n",
      "0.0\n",
      "0.0\n",
      "0.0\n",
      "0.050971632\n",
      "0.0\n",
      "1.376429483\n",
      "0.0\n",
      "4.690820074\n",
      "0.0\n",
      "0.365583678\n",
      "0.02\n",
      "0.24\n",
      "0.0\n",
      "15.7861\n",
      "0.0236756\n",
      "2.496095\n",
      "0.0\n",
      "10.56356072\n",
      "0.0\n",
      "4.24862846\n",
      "0.0\n",
      "0.024176567\n",
      "0.0\n",
      "0.050734326\n",
      "0.01274757\n",
      "0.081974113\n",
      "0.0\n",
      "1.27\n",
      "0.0\n",
      "1.134199266\n",
      "0.0\n",
      "0.856619902\n",
      "0.016571895\n",
      "0.048439408\n",
      "0.0155365\n",
      "0.309980333\n",
      "0.0\n",
      "1.580358098\n",
      "0.0\n",
      "0.315283107\n",
      "0.0\n",
      "0.114436452\n",
      "0.06\n",
      "0.14\n",
      "0.0\n",
      "0.0353662\n",
      "0.0\n",
      "0.0273516\n",
      "0.0\n",
      "0.32623474\n",
      "0.012431567\n",
      "0.333089667\n",
      "0.0\n",
      "0.234712806\n",
      "0.0\n",
      "2.573420279\n",
      "0.0\n",
      "0.46824598\n",
      "0.0\n",
      "0.381742026\n",
      "0.0\n",
      "0.562619667\n",
      "0.0\n",
      "0.628279033\n",
      "0.007833925\n",
      "0.192346978\n",
      "0.0\n",
      "0.301874047\n",
      "0.0\n",
      "1.016079447\n",
      "0.0\n",
      "0.44\n",
      "0.0\n",
      "0.121650216\n",
      "0.007293576\n",
      "2.238496734\n",
      "0.0\n",
      "0.258939159\n",
      "0.0\n",
      "0.138105761\n",
      "0.0\n",
      "0.029899648\n",
      "0.006853662\n",
      "0.084139249\n",
      "0.0\n",
      "0.35\n",
      "0.0\n",
      "0.59\n",
      "0.017470246\n",
      "0.142982769\n",
      "0.007358166\n",
      "6.47815288\n",
      "0.007322028\n",
      "0.282508787\n",
      "0.0\n",
      "0.059509027\n",
      "0.0\n",
      "0.10975815\n",
      "0.02\n",
      "0.33\n",
      "0.0\n",
      "2.114915\n",
      "0.0\n",
      "0.083242644\n",
      "0.0\n",
      "0.402643211\n",
      "0.0\n",
      "0.353097528\n",
      "0.0\n",
      "0.167497615\n",
      "0.0\n",
      "0.508396529\n",
      "165.32825756072998\n",
      "Parralel process computed in: 2.755479653676351 min\n"
     ]
    }
   ],
   "source": [
    "GE_tab_test = pd.DataFrame(index = list_genes[0:2], columns = list(set(df_ge.sample_id)))\n",
    "curr_time = time.time()\n",
    "for samp in list(set(df_ge.sample_id)):\n",
    "    test = set_ge_df_value_col(samp, df_ge, list_genes[0:2])\n",
    "print(time.time()-curr_time)\n",
    "print(f'Parralel process computed in: {(time.time()-curr_time)/60} min')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fbd13cc",
   "metadata": {},
   "source": [
    "## Get gene expression tab - Parallel iterations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "021c5435",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mp.cpu_count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "766c2a68",
   "metadata": {},
   "outputs": [],
   "source": [
    "#pool = mp.Pool(processes= mp.cpu_count())\n",
    "pool = mp.Pool(processes= 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59d036b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import partial\n",
    "t = time.time()\n",
    "GE_tab = pool.map(partial(get_ge.set_ge_df_value_col, df_ge=df_ge, list_genes=list_genes), \n",
    "                [samp for samp in list(set(df_ge.sample_id))], \n",
    "                chunksize=10)\n",
    "pool.terminate()\n",
    "print(f'Parralel process computed in: {(time.time()-t)/60} min')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "621f88d7",
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>05320234B</th>\n",
       "      <th>03660196C</th>\n",
       "      <th>05320102B</th>\n",
       "      <th>Pt4-baseline</th>\n",
       "      <th>05320444B</th>\n",
       "      <th>Pt8-baseline</th>\n",
       "      <th>03660598B</th>\n",
       "      <th>04240151B</th>\n",
       "      <th>03660501B</th>\n",
       "      <th>05320446B</th>\n",
       "      <th>...</th>\n",
       "      <th>22A</th>\n",
       "      <th>05320384B</th>\n",
       "      <th>03660555B</th>\n",
       "      <th>05420169C</th>\n",
       "      <th>05320003B</th>\n",
       "      <th>05420159C</th>\n",
       "      <th>05320216B</th>\n",
       "      <th>05320459B</th>\n",
       "      <th>05420199C</th>\n",
       "      <th>05320385B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KAT5</th>\n",
       "      <td>19.739645</td>\n",
       "      <td>27.128648</td>\n",
       "      <td>12.539045</td>\n",
       "      <td>10.42225</td>\n",
       "      <td>16.242073</td>\n",
       "      <td>19.29555</td>\n",
       "      <td>26.425709</td>\n",
       "      <td>22.062084</td>\n",
       "      <td>19.457436</td>\n",
       "      <td>17.235895</td>\n",
       "      <td>...</td>\n",
       "      <td>32.34</td>\n",
       "      <td>14.58108</td>\n",
       "      <td>11.388035</td>\n",
       "      <td>17.090183</td>\n",
       "      <td>12.617297</td>\n",
       "      <td>16.12212</td>\n",
       "      <td>6.543853</td>\n",
       "      <td>12.887998</td>\n",
       "      <td>38.806821</td>\n",
       "      <td>20.574456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4C15</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DEFB1</th>\n",
       "      <td>1.419030</td>\n",
       "      <td>0.097449</td>\n",
       "      <td>0.042814</td>\n",
       "      <td>25.54325</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>140.48550</td>\n",
       "      <td>0.048933</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.968038</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.34</td>\n",
       "      <td>2.61053</td>\n",
       "      <td>0.069665</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.205220</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.571171</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.412324</td>\n",
       "      <td>6.081646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSSC2</th>\n",
       "      <td>1.828737</td>\n",
       "      <td>3.085323</td>\n",
       "      <td>0.281840</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.146409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.372863</td>\n",
       "      <td>0.392965</td>\n",
       "      <td>1.665028</td>\n",
       "      <td>0.593880</td>\n",
       "      <td>...</td>\n",
       "      <td>3.08</td>\n",
       "      <td>0.25051</td>\n",
       "      <td>0.349408</td>\n",
       "      <td>2.490883</td>\n",
       "      <td>0.205272</td>\n",
       "      <td>7.73582</td>\n",
       "      <td>0.015569</td>\n",
       "      <td>3.198403</td>\n",
       "      <td>1.599501</td>\n",
       "      <td>0.767749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTY7</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 98 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        05320234B  03660196C  05320102B  Pt4-baseline  05320444B  \\\n",
       "KAT5    19.739645  27.128648  12.539045      10.42225  16.242073   \n",
       "OR4C15   0.000000   0.000000   0.000000       0.00000   0.000000   \n",
       "DEFB1    1.419030   0.097449   0.042814      25.54325   0.000000   \n",
       "TSSC2    1.828737   3.085323   0.281840           NaN   1.146409   \n",
       "TTTY7    0.000000   0.000000   0.000000       0.00000   0.000000   \n",
       "\n",
       "        Pt8-baseline  03660598B  04240151B  03660501B  05320446B  ...    22A  \\\n",
       "KAT5        19.29555  26.425709  22.062084  19.457436  17.235895  ...  32.34   \n",
       "OR4C15       0.00000   0.000000   0.000000   0.000000   0.000000  ...   0.00   \n",
       "DEFB1      140.48550   0.048933   0.000000   3.968038   0.000000  ...   0.34   \n",
       "TSSC2            NaN   1.372863   0.392965   1.665028   0.593880  ...   3.08   \n",
       "TTTY7        0.00000   0.000000   0.000000   0.000000   0.000000  ...   0.00   \n",
       "\n",
       "        05320384B  03660555B  05420169C  05320003B  05420159C  05320216B  \\\n",
       "KAT5     14.58108  11.388035  17.090183  12.617297   16.12212   6.543853   \n",
       "OR4C15    0.00000   0.000000   0.000000   0.000000    0.00000   0.000000   \n",
       "DEFB1     2.61053   0.069665   0.000000   3.205220    0.00000   0.571171   \n",
       "TSSC2     0.25051   0.349408   2.490883   0.205272    7.73582   0.015569   \n",
       "TTTY7     0.00000   0.000000   0.000000   0.000000    0.00000   0.000000   \n",
       "\n",
       "        05320459B  05420199C  05320385B  \n",
       "KAT5    12.887998  38.806821  20.574456  \n",
       "OR4C15   0.000000   0.000000   0.000000  \n",
       "DEFB1    0.000000   0.412324   6.081646  \n",
       "TSSC2    3.198403   1.599501   0.767749  \n",
       "TTTY7    0.000000   0.000000   0.000000  \n",
       "\n",
       "[5 rows x 98 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Concatenate all chunks to the final result\n",
    "GE_tab = pd.concat(GE_tab, axis=1)\n",
    "GE_tab.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "493f7322",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Register GEx table as csv\n",
    "GE_tab.to_csv('./melanodb_ge.csv', sep=',', header=True, index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "157c536d",
   "metadata": {},
   "source": [
    "## Get sample information table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a69a7d5c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\SARAH~1.DAN\\AppData\\Local\\Temp/ipykernel_48332/790285420.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  GE_melanoDB_info['source'] = GE_melanoDB_info['source'].replace([\"{'title': 'Genomic Features of Exceptional Response in Vemurafenib + Cobimetinib–treated Patients with BRAFV600-mutated Metastatic Melanoma', 'author': 'Yibing Yan, Antoni Ribas', 'journal': 'Clinical Cancer Research', 'location': 'San Francisco, California (United States)', 'date': 2019}\", \"{'title': 'Co-clinical assessment identifies patterns of BRAF inhibitor resistance in melanoma', 'author': 'Lawrence N. Kwong, Lynda Chin', 'journal': 'Journal of Clinical Investigations ', 'location': 'Boston (United States)', 'date': 2015}\", \"{'title': 'Acquired resistance and clonal evolution in melanoma during BRAF inhibitor therapy', 'author': 'Hubing Shi, Roger S. Lo', 'journal': 'Cancers Discovery', 'location': 'Los Angeles (United State)', 'date': 2014}\"], [1,2,3])\n"
     ]
    }
   ],
   "source": [
    "# Register table with meta informations\n",
    "GE_melanoDB_info = df_ge[['sample_id', 'source', 'temporality']]\n",
    "GE_melanoDB_info['source'] = GE_melanoDB_info['source'].replace([\"{'title': 'Genomic Features of Exceptional Response in Vemurafenib + Cobimetinib–treated Patients with BRAFV600-mutated Metastatic Melanoma', 'author': 'Yibing Yan, Antoni Ribas', 'journal': 'Clinical Cancer Research', 'location': 'San Francisco, California (United States)', 'date': 2019}\", \"{'title': 'Co-clinical assessment identifies patterns of BRAF inhibitor resistance in melanoma', 'author': 'Lawrence N. Kwong, Lynda Chin', 'journal': 'Journal of Clinical Investigations ', 'location': 'Boston (United States)', 'date': 2015}\", \"{'title': 'Acquired resistance and clonal evolution in melanoma during BRAF inhibitor therapy', 'author': 'Hubing Shi, Roger S. Lo', 'journal': 'Cancers Discovery', 'location': 'Los Angeles (United State)', 'date': 2014}\"], [1,2,3])\n",
    "GE_melanoDB_info = GE_melanoDB_info.drop_duplicates()\n",
    "GE_melanoDB_info = GE_melanoDB_info.rename(columns = {\"sample_id\": \"samplename\",\"source\": \"Batch\", \"temporality\": \"category\"})\n",
    "GE_melanoDB_info.insert(0, 'Arrayname', 'melanoDB')\n",
    "GE_melanoDB_info.to_csv('./melanodb_ge_info.csv', sep=',', header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "611f96a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Arrayname</th>\n",
       "      <th>samplename</th>\n",
       "      <th>Batch</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>melanoDB</td>\n",
       "      <td>03660196C</td>\n",
       "      <td>1</td>\n",
       "      <td>pre treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30728</th>\n",
       "      <td>melanoDB</td>\n",
       "      <td>03660356C</td>\n",
       "      <td>1</td>\n",
       "      <td>pre treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61455</th>\n",
       "      <td>melanoDB</td>\n",
       "      <td>03660445B</td>\n",
       "      <td>1</td>\n",
       "      <td>pre treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92182</th>\n",
       "      <td>melanoDB</td>\n",
       "      <td>03660447B</td>\n",
       "      <td>1</td>\n",
       "      <td>pre treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122909</th>\n",
       "      <td>melanoDB</td>\n",
       "      <td>03660501B</td>\n",
       "      <td>1</td>\n",
       "      <td>pre treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2835612</th>\n",
       "      <td>melanoDB</td>\n",
       "      <td>Pt9-baseline</td>\n",
       "      <td>{'title': 'Acquired resistance and clonal evol...</td>\n",
       "      <td>pre treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2860880</th>\n",
       "      <td>melanoDB</td>\n",
       "      <td>Pt10-baseline</td>\n",
       "      <td>{'title': 'Acquired resistance and clonal evol...</td>\n",
       "      <td>pre treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2886148</th>\n",
       "      <td>melanoDB</td>\n",
       "      <td>Pt15-baseline</td>\n",
       "      <td>{'title': 'Acquired resistance and clonal evol...</td>\n",
       "      <td>pre treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2911416</th>\n",
       "      <td>melanoDB</td>\n",
       "      <td>Pt16-baseline</td>\n",
       "      <td>{'title': 'Acquired resistance and clonal evol...</td>\n",
       "      <td>pre treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2936684</th>\n",
       "      <td>melanoDB</td>\n",
       "      <td>Pt17-baseline</td>\n",
       "      <td>{'title': 'Acquired resistance and clonal evol...</td>\n",
       "      <td>pre treatment</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>98 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Arrayname     samplename  \\\n",
       "1        melanoDB      03660196C   \n",
       "30728    melanoDB      03660356C   \n",
       "61455    melanoDB      03660445B   \n",
       "92182    melanoDB      03660447B   \n",
       "122909   melanoDB      03660501B   \n",
       "...           ...            ...   \n",
       "2835612  melanoDB   Pt9-baseline   \n",
       "2860880  melanoDB  Pt10-baseline   \n",
       "2886148  melanoDB  Pt15-baseline   \n",
       "2911416  melanoDB  Pt16-baseline   \n",
       "2936684  melanoDB  Pt17-baseline   \n",
       "\n",
       "                                                     Batch       category  \n",
       "1                                                        1  pre treatment  \n",
       "30728                                                    1  pre treatment  \n",
       "61455                                                    1  pre treatment  \n",
       "92182                                                    1  pre treatment  \n",
       "122909                                                   1  pre treatment  \n",
       "...                                                    ...            ...  \n",
       "2835612  {'title': 'Acquired resistance and clonal evol...  pre treatment  \n",
       "2860880  {'title': 'Acquired resistance and clonal evol...  pre treatment  \n",
       "2886148  {'title': 'Acquired resistance and clonal evol...  pre treatment  \n",
       "2911416  {'title': 'Acquired resistance and clonal evol...  pre treatment  \n",
       "2936684  {'title': 'Acquired resistance and clonal evol...  pre treatment  \n",
       "\n",
       "[98 rows x 4 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GE_melanoDB_info"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
